### PR TITLE
Add network layer 

### PR DESCRIPTION
**Pull request by Shiney-Kelly**
## Work Summary
- Add anime service
- Add repository and gallery viewmodel
- Add error handling
- Add get anime data models

### CODE REVIEW

File: app/src/main/java/com/project/ianime/api/AcceptLanguageInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.project.ianime.api
+
+import okhttp3.Interceptor
+import okhttp3.Response
+import java.util.*
+
+class AcceptLanguageInterceptor : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val acceptLanguage = Locale.getDefault().language
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/AcceptLanguageInterceptor.kt:10]: just to confirm, is this setting from app or from system?

File: app/src/main/java/com/project/ianime/api/error/ErrorType.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.project.ianime.api.error
+
+import androidx.annotation.StringRes
+import com.project.ianime.R
+
+enum class ErrorType(@StringRes val label: Int, @StringRes val message: Int) {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/error/ErrorType.kt:6]: need `Bad request` (400), `Content too large` (413)

File: app/src/main/java/com/project/ianime/api/error/ErrorType.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    UNAUTHORIZED(R.string.generic_error_label, R.string.unauthorized_error_message),
+    NOT_FOUND(R.string.generic_error_label, R.string.not_found_error_message),
+    CONNECTION(R.string.connection_error_label, R.string.connection_error_message),
+    GENERIC(R.string.generic_error_label, R.string.generic_error_message);
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/error/ErrorType.kt:12]: does this include `Internal error` (500)?

File: app/src/main/java/com/project/ianime/api/error/HttpException.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.project.ianime.api
+
+/**
+ * Handles specific exception such as 401, 404 and 5xx errors
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/error/HttpException.kt]: ok, so it's `ConnectionException` for handling 500 error

File: app/src/main/java/com/project/ianime/api/error/HttpStatus.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.project.ianime.api.error
+
+object HttpStatus {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/error/HttpStatus.kt:3]: 400 and 413

File: app/src/main/java/com/project/ianime/api/model/AnimeListApiModel.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.project.ianime.api.model
+
+import com.google.gson.annotations.SerializedName
+
+data class AnimeListApiModel(
+    @SerializedName("anime_list") val animeList: List<AnimeApiModel>
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/model/AnimeListApiModel.kt:9]: is this for batch get?

File: app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import com.google.gson.annotations.SerializedName
+
+/**
+ * API response of send new anime details (exclude image from a unique API)
+ */
+data class AnimeRequestApiModel(
+    @SerializedName("anime_en") val animeEnglishName: String,
+    @SerializedName("anime_zh") val animeChineseName: String,
+    @SerializedName("rate") val animeRate: Float,
+    @SerializedName("state") val animeState: String,
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt]: status

File: app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    @SerializedName("country") val animeCountry: String,
+    @SerializedName("type") val animeType: String,
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt]: genre

File: app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    @SerializedName("published_year") val animePublishedYear: String?,
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt]: release_year

File: app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    @SerializedName("description") val animeDescription: String?
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/model/AnimeRequestApiModel.kt]: synopsis

File: app/src/main/java/com/project/ianime/di/AnimeModule.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Provides a Retrofit instance
+     */
+    @Singleton
+    @Provides
+    fun providesRetrofit(okHttpClient: OkHttpClient): Retrofit {
+        return Retrofit.Builder()
+            // TODO 08-16: Add in properties build file
+            .baseUrl("https://ai8454431.pythonanywhere.com/")
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/di/AnimeModule.kt:35]: update it later

File: app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                            return@onErrorResumeNext Single.error(
+                                UnauthorizedException(it.message())
+                            )
+                        }
+                        HttpStatus.NOT_FOUND -> {
+                            return@onErrorResumeNext Single.error(
+                                NotFoundException(it.message())
+                            )
+                        }
+                        in 400..599 -> {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt]: need to add more error types?

File: app/src/main/java/com/project/ianime/screens/manageanime/ManageAnimeFragment.kt [hunk 13]
```diff
@@ -110,7 +110,7 @@ abstract class ManageAnimeFragment : BaseFragment() {
         }
     }
 
-    private fun loadInitData(){
+    private fun loadInitData() {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/screens/manageanime/ManageAnimeFragment.kt:113]: I'd use a different init setting here 😆

File: app/src/main/java/com/project/ianime/screens/manageanime/ManageAnimeFragment.kt [hunk 14]
```diff
@@ -186,7 +186,7 @@ abstract class ManageAnimeFragment : BaseFragment() {
         _binding = null
     }
 
-    companion object{
+    companion object {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/screens/manageanime/ManageAnimeFragment.kt:189]: using jpg and png right now, can support bmp, but bmp file could be large (file size limit is 1mb for now)

File: app/src/main/java/com/project/ianime/utils/image/ImageStub.kt [hunk 15]
```diff
@@ -10,7 +10,7 @@ class ImageStub {
      * Get encoded image bitmap with Base-64
      * @param [bitmap] - bitmap of current upload image
      */
-    fun getEncodedSelectedImageBitmap(bitmap: Bitmap): String{
+    fun getEncodedSelectedImageBitmap(bitmap: Bitmap): String {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/utils/image/ImageStub.kt:13]: it's no longer a base64 str, it's binary file now

File: app/src/main/res/values-zh/strings.xml [hunk 16]
```diff
@@ -54,4 +54,12 @@
...
     <string name="negative_button_label">确认</string>
     <string name="test_anime_country">国漫</string>
+
+    <!-- Error Messages  -->
+    <string name="generic_error_label">出错了</string>
+    <string name="generic_error_message">请稍后再试。</string>
+    <string name="connection_error_label">网络开小差了</string>
+    <string name="connection_error_message">您的网络信号不稳定。请稍后再试。</string>
+    <string name="unauthorized_error_message">您的权限不足以获取信息。请稍后再试。</string>
+    <string name="not_found_error_message">抱歉，没有找到任何动漫信息。请稍后再试。</string>
```
Comment (Tony-Feng) [inline app/src/main/res/values-zh/strings.xml]: maybe use comma, like `您的网络信号不稳定，请稍后再试。`

File: app/src/main/res/values/strings.xml [hunk 17]
```diff
@@ -54,4 +54,12 @@
...
     <string name="negative_button_label">Yes</string>
     <string name="test_anime_country">China</string>
+    
+    <!-- Error Messages  -->
+    <string name="generic_error_label">Something Went Wrong</string>
+    <string name="generic_error_message">Please try again later.</string>
+    <string name="connection_error_label">Connection Interrupted</string>
+    <string name="connection_error_message">Your were disconnected from the app. Please try again later.</string>
+    <string name="unauthorized_error_message">Something went wrong with your access. Please try again later.</string>
+    <string name="not_found_error_message">Unfortunately, the app could\'t find any anime information. Please try again later.</string>
```
Comment (Tony-Feng) [inline app/src/main/res/values/strings.xml]: same here, `Your were disconnected from the app, please try again later.`

File: app/src/main/java/com/project/ianime/api/AnimeService.kt [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package com.project.ianime.api
+
+import com.project.ianime.api.model.AnimeListApiModel
+import io.reactivex.rxjava3.core.Single
+import retrofit2.http.GET
+
+interface AnimeService {
+
+    //TODO: 08-15 exact api and header for get anime
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/api/AnimeService.kt:9]: do all api calls go from here?

**Review by Tony-Feng — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/project/ianime/api/AcceptLanguageInterceptor.kt (hunk 1 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/api/AcceptLanguageInterceptor.kt:10]: app language

File: app/src/main/java/com/project/ianime/api/error/ErrorType.kt (hunk 3 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/api/error/ErrorType.kt:12]: under generic

File: app/src/main/java/com/project/ianime/utils/image/ImageStub.kt (hunk 15 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/utils/image/ImageStub.kt:13]: this file will be removed, no need

File: app/src/main/java/com/project/ianime/api/model/AnimeListApiModel.kt (hunk 6 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/api/model/AnimeListApiModel.kt:9]: this is for GET request

File: app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+                                return@onErrorResumeNext Single.error(
+                                    UnauthorizedException(exception.message())
+                                )
+                            }
+                            it == HttpStatus.NOT_FOUND -> {
+                                return@onErrorResumeNext Single.error(
+                                    NotFoundException(exception.message())
+                                )
+                            }
+                            it >= 500 -> {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt:38]: add content too large err

File: app/src/main/java/com/project/ianime/viewmodels/GalleryViewModel.kt [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+                }
+            }, {
+                when (it) {
+                    is UnauthorizedException -> {
+                        animeUiState.postValue(AnimeUiState.Error(ErrorType.UNAUTHORIZED))
+                    }
+                    is NotFoundException -> {
+                        animeUiState.postValue((AnimeUiState.Error(ErrorType.NOT_FOUND)))
+                    }
+                    is ConnectionException -> {
```
Comment (Tony-Feng) [inline app/src/main/java/com/project/ianime/viewmodels/GalleryViewModel.kt:41]: add content too large err

File: app/src/main/res/values-zh/strings.xml [hunk 21]
```diff
@@ -54,4 +54,14 @@
...
     <string name="test_anime_country">国漫</string>
+
+    <!-- Error Messages  -->
+    <string name="generic_error_label">出错了</string>
+    <string name="generic_error_message">请稍后再试。</string>
+    <string name="connection_error_label">网络开小差了</string>
+    <string name="connection_error_message">您的网络信号不稳定，请稍后再试。</string>
+    <string name="unauthorized_error_message">您的权限不足以获取信息, 请联系管理员获取权限并稍后再试。</string>
+    <string name="not_found_error_message">抱歉，没有找到任何动漫信息，请稍后再试。</string>
+    <string name="content_oversize_error_message">上传数据失败，文件超载，请确认文件大小并重试。</string>
```
Comment (Tony-Feng) [inline app/src/main/res/values-zh/strings.xml]: 文件超载 -> 文件超过大小限制 (2MB)

**Review by Tony-Feng — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt (hunk 19 above)
Comment (Shiney-Kelly) [reply app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt:38]: content too large is only for post or put request right? get request only gets the url?

Comment (Tony-Feng) [reply app/src/main/java/com/project/ianime/repository/AnimeDataRepositoryImpl.kt:38]: right, only for post, patch, put, get gets url only for now

**Review by Tony-Feng — APPROVED** [review]

**Pull request merged**